src/unix/fs.c

```diff
@@ -388,29 +388,42 @@ static ssize_t uv__fs_pathmax_size(const char* path) {
 }
 
 static ssize_t uv__fs_readlink(uv_fs_t* req) {
+  ssize_t maxlen;
   ssize_t len;
   char* buf;
+  char* newbuf;
 
-  len = uv__fs_pathmax_size(req->path);
-  buf = uv__malloc(len + 1);
+  maxlen = uv__fs_pathmax_size(req->path);
+  buf = uv__malloc(maxlen);
 
   if (buf == NULL) {
     errno = ENOMEM;
     return -1;
   }
 
 #if defined(__MVS__)
-  len = os390_readlink(req->path, buf, len);
+  len = os390_readlink(req->path, buf, maxlen);
 #else
-  len = readlink(req->path, buf, len);
+  len = readlink(req->path, buf, maxlen);
 #endif
 
-
   if (len == -1) {
     uv__free(buf);
     return -1;
   }
 
+  /* Uncommon case: resize to make room for the trailing nul byte. */
+  if (len == maxlen) {
+    newbuf = uv__realloc(buf, len + 1);
+
+    if (newbuf == NULL) {
+      uv__free(buf);
+      return -1;
```

Comment:
Shouldn't we set `errno` to `ENOMEM` here?

Reply:
Shouldn't be necessary. `realloc()` sets it and `uv_free()` is carefully written to leave errno untouched. We don't set it in many other places either.

Reply:
> Shouldn't be necessary. `realloc()` sets it

Got it. Is this also the case for `malloc()`? Looking at the manual I'm not sure and we usually set  `errno` if `uv__malloc()` returns `NULL`.

Reply:
`malloc()` always sets errno on failure (barring bugs in the allocator, of course.)

I can't find many examples where we set errno outside of src/unix/fs.c, we usually `return UV_ENOMEM` or `return UV__ERR(errno)`.

Reply:
You're right, it wasn't that usual at all: I think I only account for 3. Thanks!